Remove dependency on ST CubeMX Hal

microwavesafe opened the pull request:
Using direct register writes to remove dependency on CubeMX as per forum thread

https://forum.lvgl.io/t/synchronising-flush-with-vsync/2280/20

There are three issues that need to be resolved, but it will be easier to review code here.

1.) At the moment I have directly included the stm32f429xx.h CMSIS header, obviously this isn't the right way to do it, but I was unsure how we should select the right header. We could add extra options like LV_USE_GPU_STM32F4_DMA2D and LV_USE_GPU_STM32F7_DMA2D, but this doesn't seem like a nice option. Another way would be to specify the device as well as the GPU, but I wanted to get guidance before implementing anything.

2.) I have implemented an init function that only needs to be called once.

void lv_gpu_stm32_dma2d_init(void)

At the moment I am calling this manually as I couldn't see where to put in the library otherwise. It irks me to call it on every fill or blend as it is not required, although the performance impact is probably negligible. Every little counts though!

3.) Currently I haven't implemented the masked fill, this is unused in the code so I was unsure whether we really needed it.

Lastly I had to add a attribute define to the blend_buf variable. In my setup I am putting the display and frame buffers in the main RAM and all other variables are auto allocated in the CCMRAM. Without external RAM, this gives me the biggest display buffers I can assign to improve speed. If blend_buf is auto allocated to CCMRAM then the DMA can't access it and silently fails. So I added a LV_BLEND_BUF_ATTR define to allow to to allocate it to the RAM section in my linker script.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/lvgl/lvgl?pullRequest=1561) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/lvgl/lvgl?pullRequest=1561) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/lvgl/lvgl?pullRequest=1561) it.</sub>

embeddedt commented:
> I was unsure how we should select the right header. 

This is the pattern we use elsewhere: https://github.com/lvgl/lvgl/blob/e1af20997d93c1c0afe7fca1686bbaacfb1a5ff2/lv_conf_template.h#L95

embeddedt commented:
> At the moment I am calling this manually as I couldn't see where to put in the library otherwise. 

How about calling it in `lv_init`?

kisvegabor commented on src/lv_gpu/lv_gpu_stm32_dma2d.c:
```diff
@@ -11,19 +11,14 @@
-#else
-    #error "Not supported STM32 family to use DMA2D"
-#endif
+#include "stm32f429xx.h"
```
Can't we simply replace the includes above with `stm32f7xx.h` and `stm32f4xx.h`?

microwavesafe replied:
Problem is STM32F4 and STM32F7 aren't defined unless you are using the CubeMX HAL

kisvegabor commented:
>> At the moment I am calling this manually as I couldn't see where to put in the library otherwise.

>How about calling it in lv_init?

I agree.

> So I added a LV_BLEND_BUF_ATTR define to allow to to allocate it to the RAM section in my linker script.

It seems too specific to make it a general config option. Would it help if `blend_buf` would be allocated by `lv_mem_alloc`?

embeddedt commented:
> It seems too specific to make it a general config option.

What if we generalized it to `LV_DMA_ATTR` - this variable should be placed in memory accessible by DMA?

microwavesafe commented:
> > > At the moment I am calling this manually as I couldn't see where to put in the library otherwise.
> 
> > How about calling it in lv_init?
> 
> I agree.
> 
> > So I added a LV_BLEND_BUF_ATTR define to allow to to allocate it to the RAM section in my linker script.
> 
> It seems too specific to make it a general config option. Would it help if `blend_buf` would be allocated by `lv_mem_alloc`?

No lv_mem_alloc is using space in the CCMRAM so it's not accessible from DMA. I like the LV_DMA_ATTR idea though.

microwavesafe commented:
I've changed the LV_BLEND_BUF_ATTR to LV_DMA_ATTR. I've added a define to include the correct CMSIS header. Not sure if this is what we want. And I've added the init to lv_init.

embeddedt commented on src/lv_draw/lv_draw_blend.c:
```diff
@@ -352,7 +352,8 @@ LV_ATTRIBUTE_FAST_MEM static void fill_normal(const lv_area_t * disp_area, lv_co
             if(lv_area_get_size(draw_area) >= 240) {
-                static lv_color_t blend_buf[LV_HOR_RES_MAX] = {0};
+                /* blend_buf MUST be in DMA accessible RAM (depending on linker this isn't always the case - CCMRAM) */
+                LV_DMA_ATTR static lv_color_t blend_buf[LV_HOR_RES_MAX] = {0};
```
Now that I look at it, do we need separate `blend_buf` variables for these two conditions? What do you think about moving this to the top of the file?

microwavesafe replied:
I can move it to top of file no problem.

kisvegabor replied:
We need to be sure that it's not declared if not used. On an 800x480 32 bit screen, it could mean extra 3200 byte RAM.

microwavesafe replied:
It can be wrapped in an ifdef

embeddedt commented on src/lv_gpu/lv_gpu_stm32_dma2d.c:
```diff
@@ -64,6 +56,21 @@ static DMA2D_HandleTypeDef hdma2d;
+void lv_gpu_stm32_dma2d_init(void)
+{
+    /* Enable DMA2D clock */
+    RCC->AHB1ENR |= RCC_AHB1ENR_DMA2DEN;
```
Is a `DMB` needed after this on Cortex-M7? Not 100% sure.

microwavesafe replied:
Not sure what you mean by DMB?

embeddedt replied:
It's a [data memory barrier](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.faqs/ka14041.html). That page is for ARM in general but I know the instruction exists and is sometimes needed on Cortex-M7.

microwavesafe replied:
Got you. I don't think it does as that area of memory is type Device which guarantees order.

In table B3-1 of the ARMv7-M manual (https://static.docs.arm.com/ddi0403/eb/DDI0403E_B_armv7m_arm.pdf), it states that memory block 0x40000000-0x5FFFFFFF is memory type Device. In http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0646a/CIHGEIID.html it states.

Device and Strongly-ordered
    The processor preserves transaction order relative to other transactions to Device or Strongly-ordered memory.

As those two lines of code produce 1 register read, followed by a register write, then register read all to Device type memory, the order is guaranteed.

embeddedt commented:
It looks like we need `LV_DMA_ATTR` to be added to `lv_conf_template.h` and possibly `src/lv_conf_checker.h` as well (that's why the checks are failing).

kisvegabor commented:
I suggest naming `LV_DMA_ATTR` as `LV_ATTRIBUTE_DMA` to make it similar to other [attribute defines](https://github.com/lvgl/lvgl/blob/df08d0bd543ae78a4f4b49f8c3a23199a37f6521/lv_conf_template.h#L223).

It might be worth to create `lv_gpu_stm32_dma2d_fill_opa`function and move `blend_buf` there.

microwavesafe commented:
blend_buf is also used in the general GPU callback. Here the buffer will likely need placing in memory that the DMA can access, so I think the define needs to be applied to both places.

microwavesafe commented:
I've made some changes, please let me know if you need me to change anything else. I have signed the license a couple of times, but it is still showing up as pending. Not sure why.

embeddedt commented:
> I have signed the license a couple of times, but it is still showing up as pending. Not sure why.

It's because your commits are attributed to https://github.com/springltd. The bot works by checking whether the commit authors have signed the agreement; thus that account has to sign it for the check to go away.

microwavesafe commented:
Ah OK. I have never really found a simple and quick solution for a single machine with multiple accounts, so usually I just leave it.

kisvegabor commented:
Hi,

I've tested it and works well.
For `LV_GPU_DMA2D_CMSIS_INCLUDE` I set `"stm32f769xx.h"`. Can we add some typical includes to the comment in `lv_conf_template.h` to make it more obvious for the users?

microwavesafe commented:
I've added examples.

At the moment lv_gpu_stm32_dma2d_fill_mask is wrapped in a false ifdef. Do you want me to clean this up? We could leave it so we have a reference to what it's previous function was, or I can remove all the code in the ifdef and leave a note saying not currently implemented?

kisvegabor commented:
> I've added examples.

Thanks!

> At the moment lv_gpu_stm32_dma2d_fill_mask is wrapped in a false ifdef. Do you want me to clean this up? We could leave it so we have a reference to what it's previous function was, or I can remove all the code in the ifdef and leave a note saying not currently implemented?

I think we can leave it as a reference and to make experiments with it later. 

---

All seems clear to me. As the user needs to do adjust the include file we can't add it to a patch release. So I merge it into the `dev` branch which will be merged to `master` on 16.06.
(See the updated release policy [here](https://github.com/lvgl/lvgl#release-policy))
